hphp/runtime/vm/jit/vasm-arm.cpp

```diff
@@ -1197,45 +1232,6 @@ Y(storew, m)
 
 #undef Y
 
-#define ISAR vixl::Assembler::IsImmArithmetic(value)
-#define ISLG(w) vixl::Assembler::IsImmLogical(value, w)
-#define U2(v0, v1) v0, v1
-
-#define Y(vasm_opc, lower_opc, load_opc, chk, imm, use)     \
-void lower(const VLS& e, vasm_opc& i, Vlabel b, size_t z) { \
-  lower_impl(e.unit, b, z, [&] (Vout& v) {                  \
-    auto value = safe_cast<int64_t>(i.s0.q());              \
-    if (chk) {                                              \
-      v << i;                                               \
-    } else {                                                \
-      auto s0 = v.makeReg();                                \
-      v << load_opc{imm, s0};                               \
-      v << lower_opc{s0, use, i.sf};                        \
-    }                                                       \
-  });                                                       \
-}
-
-Y(addli, addl, ldimml, ISAR, i.s0, U2(i.s1, i.d))
-Y(addqi, addq, ldimmq, ISAR, Immed64(value), U2(i.s1, i.d))
-Y(andbi, andb, ldimmb, ISLG(32), i.s0, U2(i.s1, i.d))
-Y(andli, andl, ldimml, ISLG(32), i.s0, U2(i.s1, i.d))
-Y(andqi, andq, ldimmq, ISLG(64), Immed64(value), U2(i.s1, i.d))
-Y(cmpli, cmpl, ldimml, ISAR, i.s0, i.s1)
-Y(cmpqi, cmpq, ldimmq, ISAR, Immed64(value), i.s1)
-Y(orqi, orq, ldimmq, ISLG(64), Immed64(value), U2(i.s1, i.d))
-Y(subli, subl, ldimml, ISAR, i.s0, U2(i.s1, i.d))
-Y(subqi, subq, ldimmq, ISAR, Immed64(value), U2(i.s1, i.d))
```

Comment:
I basically agree with your conclusion at the end—I think we should go through vasm-lower for specifically those instructions which need to be lowered (e.g., because they would otherwise require scratch registers) or because doing so is an optimization (e.g., the `movtq*{}` instructions, which otherwise copy-propagation couldn't eliminate).  I definitely don't think the distinction should be tied to "what exists" on the architecture.